### PR TITLE
Output single grid DEM

### PR DESCRIPTION
**Pull request by voferreira**
# Description of the problem

- In DEM, the grid is output in every single output step;

# Description of the solution

- Create a member function for the grid output separated from the output of the particles data
- Move the member function to Lagrangian post-processing

# How Has This Been Tested?

- Sedimentation of multiple particles case - All scenarios

# Documentation

- Created post-processing parameters guide

### CODE REVIEW

File: include/dem/dem.h
Comment (blaisb) [inline include/dem/dem.h:239]: What is an output grid? is it the boundary? the grid of the domain?

**Review by blaisb — CHANGES_REQUESTED** [review]
Can you just comment a bit better in the .h? Otherwise this should be ok

**Comment by blaisb** [comment]
From my understanding, this prevents us from also outputting the grid to monitor how to load balancing is proceeding no? I feel that's a bad thing, because it is very important for many applications to see the load balancing progress... Do you think there's a possible compromise somewhere?

**Comment by voferreira** [comment]
> From my understanding, this prevents us from also outputting the grid to monitor how to load balancing is proceeding no? I feel that's a bad thing, because it is very important for many applications to see the load balancing progress... Do you think there's a possible compromise somewhere?

You are right. I didn't think about that when I did it. I thought of it and I think the best way to address this is to add a parameter to control this, and call ``write_output_grid()`` within ``write_output_results()`` if this parameter is set to ``true``. What do you think?

File: source/dem/dem.cc
Comment (voferreira) [inline source/dem/dem.cc:682]:
```suggestion
```

File: source/dem/dem.cc
Comment (blaisb) [inline source/dem/dem.cc:630]: Do you think there is a way that this grid could be merged with the lagrangian post-processing or you think this is too difficult? E.g. if we have lagrangian post-processing the info would just be added on top of this grid

**Review by blaisb — CHANGES_REQUESTED** [review]
Just a question, otherwise seems good to me.

File: doc/source/parameters/dem/post-processing.rst
Comment (blaisb) [inline doc/source/parameters/dem/post-processing.rst:62]: So I was wondering, do you think there is a way this data could be added to the grid files directly instead of being appended to different variables? E.g. make it a single file with all the possible lagrangian post processing capabilities instead of a single one? Or is it already like this? :). I know i'm asking a lot but this would be very useful.

**Review by blaisb — APPROVED** [review]
Very cool PR. I am thinking that in the future all Lagrangian post-processing files could merged into the grid written. Maybe not for this PR this is already a lot :).

**Comment by blaisb** [comment]
@voferreira can you get a second review from someone else? Once that's done I'll merge

File: doc/source/parameters/dem/post-processing.rst
Comment (voferreira) [reply doc/source/parameters/dem/post-processing.rst:62]: No worries, it is not asking a lot. I am trying to do it. I will push it later today if I manage to do so. I think it is the best approach in order to clean a bit the solver.

**Comment by voferreira** [comment]
I just refactored the postprocessing in order to create a single file for the granular temperature, the particles' velocity, and the grid. I like the new pattern because the dem.cc and the cfd_dem_coupling.cc files get cleaner, additionally, we have a single file for all grid-related data.

**Comment by blaisb** [comment]
> I just refactored the postprocessing in order to create a single file for the granular temperature, the particles' velocity, and the grid. I like the new pattern because the dem.cc and the cfd_dem_coupling.cc files get cleaner, additionally, we have a single file for all grid-related data.

You are getting real good at this :)! Thanks, life will be much cleaner now

**Review by blaisb — APPROVED** [review]
Good job! Very cool. This is so much cleaner than it used to be.

File: doc/source/parameters/dem/post-processing.rst
Comment (acdaigneault) [inline doc/source/parameters/dem/post-processing.rst:36]: Can you explain a bit what disabling `write grid` does? For my understanding after the conversation in this PR, enabling write grid will output grid as before the changes, but I'm not sure what happen when it's false?

File: doc/source/parameters/dem/post-processing.rst
Comment (acdaigneault) [inline doc/source/parameters/dem/post-processing.rst:50]: It is possible to set the default value the same as the time end in the Simulation control subsection?

Comment (acdaigneault) [reply doc/source/parameters/dem/post-processing.rst:50]: I just saw that this is the step and not the time. Is there's a reason why? I feel the time is easier to use.

File: doc/source/parameters/dem/post-processing.rst
Comment (acdaigneault) [inline doc/source/parameters/dem/post-processing.rst:38]: Can you explain a bit how the average velocity is calculated and how it is represented in the output? I don't understand how it works since it's exported on a grid.

File: doc/source/parameters/dem/post-processing.rst
Comment (acdaigneault) [reply doc/source/parameters/dem/post-processing.rst:36]: I tried the parameter and it generates `-postprocess_data` files and not independent `-grid` files.

**Review by acdaigneault — COMMENTED** [review]
Hey nice PR! I have a comment that I did not address here which may lead to a discussion.
I'm not sure that I like the grid output with the post-process with the current frequency/start step/end step default values. It seems a lot of thinking to set proper parameters only to have a grid visualization. I think that's very interesting to have access to parameters to help on the grid output. But, in my opinion, I think it would be nice to have something similar as before if `write grid` is enable without having to set the other parameters. What do you think?

**Comment by voferreira** [comment]
> Hey nice PR! I have a comment that I did not address here which may lead to a discussion. I'm not sure that I like the grid output with the post-process with the current frequency/start step/end step default values. It seems a lot of thinking to set proper parameters only to have a grid visualization. I think that's very interesting to have access to parameters to help on the grid output. But, in my opinion, I think it would be nice to have something similar as before if `write grid` is enable without having to set the other parameters. What do you think?

Thanks! You mean having the previous parameters as default? If that's it I agree with you. The current defaults need more attention from the user than necessary. I will try to address this.

File: doc/source/parameters/dem/post-processing.rst
Comment (voferreira) [reply doc/source/parameters/dem/post-processing.rst:50]: I think it is because it is an integer, not a double. Ifs with doubles are always a problem

**Comment by acdaigneault** [comment]
@voferreira Yeah for instance, let's say I want a grid from the beginning to the end of the simulation. I don't know what is the end step, I just know the time. This is why I don't like the "end step" but I understand it's better for the ifs. Is the end step relevant for the post-processing? If not, maybe it should be stopped?

**Comment by blaisb** [comment]
I think when write grid is true and the lagragian post-processing is true, then the grid should always be written.
The time start and time end are really used to decide WHEN you average the velocity and the temperature, and not when you output.
It's a very good comment @acdaigneault

**Comment by voferreira** [comment]
As discussed, I removed the excessive parameters from the code and explained better what the Lagrangian post-processing does. The idea now is that the post-processing is done altogether when ``Lagrangian post-processing = true``, and the output of the post-processing follows the output of the particles. I think it is better to keep it simple like it is now. Please @acdaigneault and @blaisb, let me know what you think of it.

**Comment by blaisb** [comment]
One thing @voferreira don't forget to update the prm file of the example also :)!

File: source/dem/lagrangian_post_processing.cc
Comment (acdaigneault) [reply source/dem/lagrangian_post_processing.cc:203]: I was thinking about the same thing

Comment (acdaigneault) [reply source/dem/lagrangian_post_processing.cc:203]: @voferreira I think storage is [avg_x_0, avg_y_0, avg_z_0, avg_x_1, ...] You can look at line 1570 of `navier_stokes_base.cc`, but not 100% sure since it's DoF data and not cell data, but it's should be similar. :)

File: doc/source/parameters/dem/post-processing.rst
Comment (acdaigneault) [inline doc/source/parameters/dem/post-processing.rst:12]: Yay! I like it

**Review by acdaigneault — APPROVED** [review]
Thanks for the changes, I like it! Doc is really useful! I gave maybe for tips for the Bruno's comment. Also, I tried it and it works well. I have to say that watching the average velocities in cells reminds me of Minecraft haha.

**Pull request merged**
